### ISSUE TITLE
[Question] How can I achieve ClickHouse synchronization between two Kubernetes clusters?

### ISSUE THREAD

**Issue by sawchen** [Altinity/clickhouse-operator#1380]
I have successfully deployed ClickHouse using clickhouse-operator-altinity in one of the Kubernetes clusters (Cluster A), utilizing clickhouse-keeper for replica synchronization. The deployment configuration is as follows:

```yaml
apiVersion: clickhouse.altinity.com/v1  
kind: ClickHouseInstallation  
metadata:  
  name: my-clickhouse  
spec:  
  defaults:  
    templates: 
      dataVolumeClaimTemplate: default-volume-claim
      logVolumeClaimTemplate: default-volume-claim 
  configuration:
    users:
      user1/networks/ip: "::/0"
      user1/password: abc098pmdd
      user1/profile: default  
    clusters:  
      - name: replicas-only  
        templates:
          podTemplate: clickhouse-v23.8
          dataVolumeClaimTemplate: default-volume-claim
          logVolumeClaimTemplate: default-volume-claim
        layout:
          # shardsCount not specified, assumed = 1, by default
          replicasCount: 3 
    zookeeper:  
      nodes:  
        - host: clickhouse-keeper.clickhouse.svc.cluster.local  
          port: 2181  
  templates:
    volumeClaimTemplates:
      - name: default-volume-claim
        provisioner: StatefulSet
        reclaimPolicy: Retain
        spec:  
          storageClassName: csi-rbd-sc  
          accessModes:  
            - ReadWriteOnce
          resources:
            requests:
              storage: 20Gi
    podTemplates:
      - name: clickhouse-v23.8
        spec:
          containers:
            - name: clickhouse
              image: clickhouse/clickhouse-server:23.8
              volumeMounts:
                - name: default-volume-claim
                  mountPath: /var/lib/clickhouse
              resources:
                requests:
                  memory: "256Mi"
                  cpu: "200m"
                limits:
                  memory: "4096Mi"
                  cpu: "2000m"
```

The current operational status is:

```
kubectl get all -n clickhouse
NAME                                                                 READY   STATUS    RESTARTS        AGE
pod/chi-my-clickhouse-replicas-only-0-0-0                            2/2     Running   0               4h35m
pod/chi-my-clickhouse-replicas-only-0-1-0                            2/2     Running   2 (3h27m ago)   4h19m
pod/chi-my-clickhouse-replicas-only-0-2-0                            2/2     Running   8 (3h27m ago)   9d
pod/clickhouse-keeper-0                                              1/1     Running   6 (3h52m ago)   15d
pod/clickhouse-keeper-1                                              1/1     Running   4 (3h58m ago)   15d
pod/clickhouse-keeper-2                                              1/1     Running   3 (3h58m ago)   15d
pod/clickhouse-operator-altinity-clickhouse-operator-6964d78-ptfxq   2/2     Running   6 (23h ago)     16d

```

Now, how should I write the deployment configuration file for my second Kubernetes cluster (Cluster B)?

Should I first install clickhouse-operator-altinity in Cluster B without deploying clickhouse-keeper? Then, referring to the ClickHouse configuration in Cluster A, should I change the zookeeper connection information to connect to the clickhouse-keeper in Cluster A, while keeping all other information, including usernames, passwords, etc., unchanged?

I have seen samples related to zones and distribution, such as:
zones-04-advanced-04-pod-per-host-local-storage.yaml
13-distribution-02-3x3-circular-replication.yaml

However, I'm not sure if these are applicable to my scenario of having two separate Kubernetes clusters, and I'm not using Amazon cloud Web Services. My Kubernetes clusters are deployed on bare metal in two separate data centers.

**Comment by Slach** [comment]
first of all, 
you need to apply something solutions which allow access clickhouse-server from cluster B to keeper in cluster A

try look to 
https://github.com/liqotech/liqo
https://github.com/submariner-io/submariner

after keeper serivce will available in cluster B

then you can try to install kind: ClickHouseInstallation
with 
```yaml
spec:
 configuration:
   clusters:
   - name: cluster-b
     shardsCount: 1
     replicasCount: 1  
   zookepeer:
```
which reference to keeper service in cluster A

use `engine=ReplicatedMergeTree('/clickhouse/tables/{shard}/{database}/{table}','{replica}')` in your table schemas
in this case your cluster-b will have same data with `replicas-only` in cluster A

**Comment by sawchen** [comment]
@Slach  I have a general understanding of the approach you mentioned.

Currently, I am able to expose the ClickHouse service in Kubernetes cluster A to the outside using MetalLB, which allows services outside the cluster to access it.

------------------------------

kubectl get svc -n clickhouse
NAME                                                       TYPE           CLUSTER-IP       EXTERNAL-IP    PORT(S)                                        AGE
clickhouse-keeper                                          LoadBalancer      10.101.42.210    10.30.60.190        2181:32234/TCP,7000:31862/TCP                              15d

------------------------------

As for Kubernetes cluster B, would it be feasible to install ClickHouse with a configuration that specifies the use of the keeper in Kubernetes cluster A, while keeping all other information unchanged?

-------------------------------

apiVersion: clickhouse.altinity.com/v1  
kind: ClickHouseInstallation  
metadata:  
  name: my-clickhouse  
spec:  
  defaults:  
    templates: 
      dataVolumeClaimTemplate: default-volume-claim
      logVolumeClaimTemplate: default-volume-claim 
  configuration:
    users:
      user1/networks/ip: "::/0"            #####  unchanged
      user1/password: 123qeoesrgt     #####  unchanged
      user1/profile: default              #####  unchanged
    clusters:  
      - name: replicas-only      #####  unchanged
        templates:
          podTemplate: clickhouse-v23.8
          dataVolumeClaimTemplate: default-volume-claim
          logVolumeClaimTemplate: default-volume-claim
        layout:
          # shardsCount not specified, assumed = 1, by default
          replicasCount: 3 
    zookeeper:  
      nodes:  
        - host: 10.30.60.190   ####keeper in cluster A 
          port: 2181  
  templates:
        ...

-------------------------

**Comment by Slach** [comment]
yes, following config section looks good
```yaml
    zookeeper:  
      nodes:  
        - host: 10.30.60.190   ####keeper in cluster A 
          port: 2181  
```
check availability for 10.30.60.190:2181

new clickhouse cluster in k8s cluster-B shall have different name with clickhouse cluster in k8s cluster-A to avoid conflicts of {replica} macros values

**Comment by sawchen** [comment]
@Slach 
Cluster A remains unchanged as per the above deployment. Now, I have linked Cluster B with Cluster A using the capabilities of cilium clustermesh. I have created a new svc for clickhouse-keeper in Cluster A as follows:
`apiVersion: v1
kind: Service
metadata:
  annotations:
    service.cilium.io/global: "true"
    service.cilium.io/shared: "true"
  labels:
    app: clickhouse-keeper
  name: clickhouse-keeper-global
  namespace: clickhouse
spec:
  ports:
  - name: client
    port: 2181
    protocol: TCP
    targetPort: 2181
  - name: prometheus
    port: 7000
    protocol: TCP
    targetPort: 7000
  selector:
    app: clickhouse-keeper
    what: node
  sessionAffinity: None
  type: ClusterIP`

At the same time, I also created the same svc in Cluster B with the same configuration. Now, I can query and access the clickhouse-keeper of Cluster A by starting the zookeeper cli in Cluster B. The information is as follows:
`[zk: clickhouse-keeper-global.clickhouse.svc.cluster.local:2181(CONNECTED) 20] ls /clickhouse

[my-clickhouse, replicas-only, table, tables]`

I want to achieve that after Cluster B starts, the clickhouse can synchronize all data with the original Cluster A.
So, what changes should be made in the clickhouse of Cluster B? Please see below:

apiVersion: clickhouse.altinity.com/v1  
kind: ClickHouseInstallation  
metadata:  
  name: my-clickhouse ###Should this naming be consistent with Cluster A？
spec:  
  defaults:  
    templates: 
      dataVolumeClaimTemplate: default-volume-claim
      logVolumeClaimTemplate: default-volume-claim 
  configuration:
    users:
      user1/networks/ip: "::/0"
      user1/password: ckadmin2024
      user1/profile: default  
    clusters:  
      - name: replicas-only  ###Should this naming be consistent with Cluster A？
        templates:
          podTemplate: clickhouse-v23.8
          dataVolumeClaimTemplate: default-volume-claim
          logVolumeClaimTemplate: default-volume-claim
        layout:
          # shardsCount not specified, assumed = 1, by default
          replicasCount: 3 
    zookeeper:  
      nodes:  
        - host: clickhouse-keeper-global.clickhouse.svc.cluster.local  
          port: 2181  
  templates:
    volumeClaimTemplates:
      - name: default-volume-claim
        provisioner: StatefulSet
        reclaimPolicy: Retain
        spec:  
          storageClassName: csi-rbd-sc  
          accessModes:  
            - ReadWriteOnce
          resources:
            requests:
              storage: 20Gi
    podTemplates:
      - name: clickhouse-v23.8
        spec:
          containers:
            - name: clickhouse
              image: clickhouse/clickhouse-server:23.8
              volumeMounts:
                - name: default-volume-claim
                  mountPath: /var/lib/clickhouse
              resources:
                requests:
                  memory: "256Mi"
                  cpu: "200m"
                limits:
                  memory: "4096Mi"
                  cpu: "2000m"

**Comment by sawchen** [comment]
@Slach 
I have successfully deployed Cluster B using different parameters from Cluster A as per the following configuration.

```
apiVersion: clickhouse.altinity.com/v1  
kind: ClickHouseInstallation  
metadata:  
  name: my-clickhouse-b  ### different name
clusters:  
      - name: replicas-only-b  ### different cluster name
```
A month ago, I created a database and data table in my Cluster A and imported one piece of data. It can run normally under Cluster A. The statements to create the database and table are as follows:

```
CREATE DATABASE test_db ON CLUSTER 'replicas-only';

CREATE TABLE IF NOT EXISTS test_db.za_history ON CLUSTER 'replicas-only' (
    za_name String,
    za_data_id UInt64,
    itemid Int64,
    create_time DateTime
) 
ENGINE = ReplicatedMergeTree('/clickhouse/tables/01/{database}/{table}', '{replica}')
PARTITION BY toYYYYMMDD(create_time)
ORDER BY create_time
```
Now, I have also created the corresponding database and data table in Cluster B, but the parameters use the name of Cluster B, as follows:
```
CREATE DATABASE test_db ON CLUSTER 'replicas-only-b';

CREATE TABLE IF NOT EXISTS test_db.za_history ON CLUSTER 'replicas-only-b' (
    za_name String,
    za_data_id UInt64,
    itemid Int64,
    create_time DateTime
) 
ENGINE = ReplicatedMergeTree('/clickhouse/tables/01/{database}/{table}', '{replica}')
PARTITION BY toYYYYMMDD(create_time)
ORDER BY create_time
```
Now, I find that the data of Cluster A has not been synchronized to Cluster B. When I query the information through the ZooKeeper CLI, there is information about Cluster B.

```
[zk: clickhouse-keeper-global.clickhouse.svc.cluster.local:2181(CONNECTED) 35]
 ls /clickhouse/tables/2/test_db/za_history/replicas
[chi-my-clickhouse-b-replicas-only-b-0-0, chi-my-clickhouse-b-replicas-only-b-0-1, chi-my-clickhouse-b-replicas-only-b-0-2, chi-my-clickhouse-replicas-only-0-0, chi-my-clickhouse-replicas-only-0-1, chi-my-clickhouse-replicas-only-0-2]
```
Is there something wrong with my configuration? Or, to trigger data synchronization, have I missed any steps, such as inserting one or two pieces of data into za_history to activate synchronization?

**Comment by sawchen** [comment]
@Slach 
Thank you for your correction, my database creation statement was wrong. The query for Cluster A found that it is using `ReplicatedMergeTree(‘/clickhouse/tables/2/{database}/{table}’, ‘{replica}’). `
So Cluster B should also use 
`ReplicatedMergeTree(‘/clickhouse/tables/2/{database}/{table}’, ‘{replica}’).`

Because my cluster does not plan to use the sharding feature, so shardsCount is set to 1, so {shard} uses a fixed value. Previously, I arbitrarily filled in 2, so I have been using 2 as the directory structure.

In this way, the node path of the same set of clickhouse-keeper is adapted. At this time, using the query statement you provided, I see the synchronization progress. 

I found that in the cluster synchronization information, the replication_queue indicates that DNS cannot resolve the host corresponding to the replica, and the connection times out. It should be that the local cluster cannot know the information across clusters.

The solution I found:
Declare the svc of my-clickhouse in Cluster A to be deployed in Cluster B Declare the svc of my-clickhouse-b in Cluster B to also be deployed in Cluster A. At the same time, add the following in these svcs: 
```
annotations: 
   service.cilium.io/global: “true” 
   service.cilium.io/shared: “true”
```
Finally, the data synchronization was successful. 
Thank you for your guidance!
